### ISSUE TITLE
Support Inputstick

### ISSUE THREAD

**Issue by github-k8n** [Kunzisoft/KeePassDX#310]
Keepass2Android supports plugins, one of which allows the use of Inputstick http://www.inputstick.com/
https://github.com/inputstick/KP2APlugin

A feature request would be to natively support Inputstick to type username/password/URL etc.
https://github.com/inputstick
https://github.com/inputstick/InputStickAPI-Android

**Comment by J-Jamet** [comment]
Linked to https://github.com/Kunzisoft/KeePassDX/issues/259 #80

**Comment by schlomie** [comment]
A modest proposal, or as a stopgap...

You wouldn't necessarily _need_ to integrate InputStick's API, at least not right away.

This came up in F-Droid not so long ago:
https://github.com/TheLastProject/ShareToInputStick
https://f-droid.org/en/packages/me.hackerchick.sharetoinputstick/

It acts as a share target, and then sends whatever arbitrary text you share with it to your InputStick.  You would just need some kind of mechanism to select whichever field (username, password, email, attribute) you wanted to share; have it bring up the "Share With..." dialog, tap on "Share To InputStick" and you're done.

That said, I'm not much of an Android developer and I don't really know what all is involved with Android's Share With API - if there are any security concerns, so on and so forth.

Just throwing this out there as an idea.

**Comment by super9mega** [comment]
I just got an input stick and would love this to be implemented (I know its not as helpful as something like actual code but I wanted to show support). I work as IT and this would help out and make my day to day operations so much faster, currently the way to do it is to copy then use "usb Remote" in order to paste it, and it just adds alot of delay I could be using to fix the issues themselves. if you need any help I'm here and can test versions with this in it as well.

**Comment by Maingron** [comment]
Yeah, I need to send my credentials to Inputstick on a daily basis.
Unfortunately, without such integration, I have to stick to Keepass2Android for now

**Comment by wschlich** [comment]
I'd also **love** to have native InputStick support directly integrated in KeePassDX! <3

> A modest proposal, or as a stopgap...
> 
> You wouldn't necessarily _need_ to integrate InputStick's API, at least not right away.
> 
> This came up in F-Droid not so long ago: https://github.com/TheLastProject/ShareToInputStick https://f-droid.org/en/packages/me.hackerchick.sharetoinputstick/
> 
> It acts as a share target, and then sends whatever arbitrary text you share with it to your InputStick. You would just need some kind of mechanism to select whichever field (username, password, email, attribute) you wanted to share; have it bring up the "Share With..." dialog, tap on "Share To InputStick" and you're done.

`ShareToInputStick` itself works kinda well for me, but many times, connecting to an InputStick fails for the first time, so you have to re-select/tap the same device in `ShareToInputStick` at least twice (update: issue [now reported](https://github.com/TheLastProject/ShareToInputStick/issues/108)). The KP2A InputStick plugin works a lot better for me in this regard (update: but also has [a minor issue](https://github.com/inputstick/KP2APlugin/issues/11)).

When considering to support `ShareToInputStick` (would be okay for a temporary workaround IMHO), please use the sharing intent `text/x-keystrokes` which was added to `ShareToInputStick` on my request [a while ago](https://github.com/TheLastProject/ShareToInputStick/issues/90). The benefit is that you can select to **always** use the `ShareToInputStick` app for this intent type, saving an extra step compared to using the `text/plain` intent type.

**Comment by wschlich** [comment]
@J-Jamet I see you have moved this issue over to your [TODO list for KPDX5](https://github.com/orgs/Kunzisoft/projects/10/views/1) at the end of last year. Thank you for considering it! How likely is it that you'll be _actually implementing_ InputStick support for KPDX5, taking into account that this issue is open for over 6 years now and thus likely a low priority for you?
Also, do you have a rough estimate on when you plan to work on KPDX5?
I suppose that it will likely take quite a long time (several months or even longer) until we could expect this feature to be available in KPDX(5). Maybe you could consider implementing the `ShareToInputStick` workaround at first (see my previous posting).
Thanks and please keep up the great work with KPDX! 🙏

**Comment by J-Jamet** [comment]
Indeed, it wasn't a high priority, even though it might be a good addition, as it requires additional hardware. I don't have any estimates; the features are created as we go along. I'd like to set it up for personal use as well, but I don't even have the hardware. I'd like to explore the possibility of direct use from a modified ROM.
KeePassDX version 5 will be a refactoring of the code, so this will be an opportunity.

**Comment by Maingron** [comment]
@J-Jamet if you happen to be located in Germany in or near Leipzig - I had bulk-ordered 4 to save future shipping costs, but only ever needed 2 of my 4 sticks - you can have one without shipping costs, if it helps development! :)

Alternatively I can also offer some testing!

**Comment by ei-ke** [comment]
@J-Jamet @Maingron 
I  had exactly the same thought when I received the notification and also would have donated an spare Inputstick laying around. If it won't be an issue for @Maingron I'd happily donate the international shipping fees to your Inputstick :)

**Comment by J-Jamet** [comment]
Thanks, that's kind of you, but I'll order it myself when I work on it. If this equipment is required and not readily available to everyone, I will focus on a solution that initially only requires a software modification.

**Comment by schlomie** [comment]
https://github.com/tejado/Authorizer is a PasswdSafe fork that implements autotype over USB and Bluetooth.  The USB HID works well, but requires root.  Bluetooth autotype is rather clunky to connect properly.  That said, it may prove helpful in working out some of the mechanics at least.

Thank you!

**Comment by Maingron** [comment]
@J-Jamet it definitely is available to everyone, but the shipping costs had been very high. It seems like they dropped now, but at least back then it made more sense to order many once instead of multiple times.

Glad things are improving on both ends! :)

**Comment by wschlich** [comment]
> Indeed, it wasn't a high priority, even though it might be a good addition, as it requires additional hardware. I don't have any estimates; the features are created as we go along. I'd like to set it up for personal use as well, but I don't even have the hardware. I'd like to explore the possibility of direct use from a modified ROM. KeePassDX version 5 will be a refactoring of the code, so this will be an opportunity.

Could you imagine adding support for sharing passwords (or even better: any other string fields' values) with the `text/x-keystrokes` intent type that I mentioned in [my comment](#issuecomment-3311404311) through an additional button (like the currently implemented copy button) or similar?

It would require much less efforts and **no hardware** for you to test (you could just check whether `KDE Connect` and/or `ShareToInputStick` appear in the app list that pops up after firing up the intent :-)

**Comment by github-k8n** [comment]
Problem is: sharetoinputstick doesn't work at all with non-US keyboard.
(I have german keyboard layout in android, laptop (linux) and in the inputstick utility as default keyboard layout)
Whether connecting directly or by sending it to inputstickutility it will always mess up and send as US keyboard
(effectively sending stuff like  "httpsÖ--github.com-"  instead of "https://github.com/" )

Using USB remote ( https://www.inputstick.com/usb-remote-android/ ) it works correctly.

So using sharetoinputstick is not a great solution unless you're an all-US-keyboard shop, or I am missing something here..

**Comment by wschlich** [comment]
> Problem is: sharetoinputstick doesn't work at all with non-US keyboard. (I have german keyboard layout in android, laptop (linux) and in the inputstick utility as default keyboard layout) Whether connecting directly or by sending it to inputstickutility it will always mess up and send as US keyboard (effectively sending stuff like "httpsÖ--github.com-" instead of "https://github.com/" )

I just created the following issue in the ShareToInputStick project (which you could have done as well, btw.) to support other keyboard layouts: https://github.com/TheLastProject/ShareToInputStick/issues/110

> So using sharetoinputstick is not a great solution unless you're an all-US-keyboard shop, or I am missing something here..

Adding support for sharing text with the `text/x-keystrokes` intent type would also make it possible to send it via the KDE Connect app (`Send as keystrokes`) which is also awesome for computers that you are allowed and/or able to install software on (it even runs on Windows).

**Comment by Maingron** [comment]
I'm not sure how secure and reliable sharing to KDE Connect would be, but having options is always nice, and I'll certainly use it one day!  

Tho we must make sure the password doesn't route through the internet, especially if the user isn't aware.

**Comment by wschlich** [comment]
> I'm not sure how secure and reliable sharing to KDE Connect would be, but having options is always nice, and I'll certainly use it one day!

You should definitely try it, it's a game changer.

> Tho we must make sure the password doesn't route through the internet, especially if the user isn't aware.

Well, that's an unapplicable scenario because KDE Connect _only_ uses a local network to which both the phone running the KDE Connect app and the computer running the KDE Connect desktop program (Linux with KDE or Windows) must be connected (WiFi or Ethernet doesn't matter).

**Comment by ishanarora** [comment]
According to [KDE Connect's Privacy Policy](https://community.kde.org/KDEConnect/PrivacyPolicy)
> KDE Connect never sends any information to KDE nor to any third party. KDE Connect sends data from one device to the other directly using the local network, never through the internet. Network traffic is always encrypted using TLS1.2.

**Comment by Maingron** [comment]
Thanks guys! There is a "network" option. I probably got it mixed up with "online".

I do use KDE connect, it definitely is a game changer! I'm just not sure if I'll share passwords using KDE Connect.

**Comment by github-k8n** [comment]
I think the topic got derailed a bit here. Let's maybe focus on Inputstick first instead of discussing KDE connect features...
If that solution (text/x-keystrokes ?) happens to also be good for stuff like that KDE connect, great.

But I think at least in this issue the focus should be with inputstick and to make it as seamless ab experience as it is with keepass2android, not focusing on making it some kind of a cludgy solution that will support ALL other methods in a suboptimal way.

**Comment by larrylart** [comment]
Ha! I didn’t know about InputStick until I saw it mentioned here last week.
I’d actually been looking for a solution like this for a while — but never came across it, so I went ahead and built my own using a $17 ESP32-S3 dongle (see #2186 ). I integrated it into a modified clone of KeePassDX, mainly as a proof of concept since I wasn’t sure if anyone else would be interested. It seems to be working well, as far as I could test it, on windows, linux, andoird and iphone. 

I know it’s not directly related to the main topic (InputStick), but I think it’s a step in the right direction to get the conversation and development moving. Under the same UI/settings, I think the InputStick SDK could also be supported in the future.
My implementation is completely open-source on the dongle side, so others can adapt or extend it however they like.

**Comment by larrylart** [comment]
Regarding InputStick integration, if anyone is interested in testing a possible direct integration via a modified InputStick KeePass2Android (KP2A) plugin:

As part of the Blue Keyboard dongle integration with KeePassDX, I’ve created a new fork of KeePassDX that implements an AIDL-based approach:
https://github.com/larrylart/KeePassDX

This keeps the KeePassDX integration minimal and clean, and also makes it much easier for other HID-style providers to integrate with KeePassDX by implementing the same AIDL service contract.

As a proof of concept, I also modified the InputStick KeePass2Android plugin (fork available at https://github.com/larrylart/KP2APlugin) so it can receive credentials from the KeePassDX AIDL fork.

Unfortunately, I don’t have an InputStick myself, so I can only confirm that the plugin receives the message; I can’t verify whether it successfully types the credentials on the target device. I blindly wired the internals based on what was originally done. 

If anyone who owns an InputStick would like to test this, the release (including all required files, such as the modified InputStick KP2A plugin APK) is available here:
https://github.com/larrylart/KeePassDX/releases

Note: Please ignore my crude UX plumbing ... a cleaner approach would likely be something like tapping a field to send, or using a top-right menu to send predefined combinations (e.g. username<tab>password<enter>). I am not the best at this task ...

![Image](https://github.com/user-attachments/assets/ab290a2b-987e-4e01-b4bd-393c3ea3548e)
![Image](https://github.com/user-attachments/assets/478085f5-8af8-464c-907d-17236c343ee9)
![Image](https://github.com/user-attachments/assets/ce0811dc-d667-4d6e-b4ee-86d01efb8f98)

**Comment by ei-ke** [comment]
@larrylart Thank you for your work. I'll test your forks with my InputStick. I hope to do so until 01.01.2026

**Comment by Maingron** [comment]
@larrylart seems to be working. See attachments. InputStickUtility 2.12 and the latest versions of your forks. Let me know if you need to know something else :)

<img width="1080" height="1511" alt="Image" src="https://github.com/user-attachments/assets/61b8831e-3dfd-456c-b93f-2f51e86119a8" />

https://github.com/user-attachments/assets/e8be852c-f25c-44be-8c8f-12f5f55e6d05

**Comment by larrylart** [comment]
> [@larrylart](https://github.com/larrylart) seems to be working. See attachments. InputStickUtility 2.12 and the latest versions of your forks. Let me know if you need to know something else :)
> 
Great! That button is really firing events like crazy! me bad UI plumbing! 😂  
Honestly I do not know where to take it from here, InputStick development/activity on github seems to be stuck two years ago. Option, if going this way/AIDL integration, finish integration for this for other send modes (use, user_tab_pass_enter, user_enter_pass_enter, etc - depending on what keepassdx interface is agreed on - it could be completely generic), but than to release it ... I doubt that  taken on the official channel. Another option could be a trimmed down version just for KeePassDX that @J-Jamet can release alongside with the other KeePassDX related apps. 

As for the UX, my "play" button, although I like to have a button :), is not really the best approach in UI give that other fields, user, maybe url, etc., might need to have the send option as well. Perhaps tap on the field itself to trigger send if provider enabled, maybe popup to confirm/send for accidental interaction. And for the combos (user_enter_pass_enter, etc) maybe have a few menu option in the top-right, only available when provider set and enabled. What do you think?
